stdlib/public/core/Sort.swift

```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
   ) rethrows {
     let didSortUnsafeBuffer = try _withUnsafeMutableBufferPointerIfSupported {
       buffer -> Void? in
-        try buffer.sort(by: areInIncreasingOrder)
+        try buffer._stableSortImpl(by: areInIncreasingOrder)
     }
     if didSortUnsafeBuffer == nil {
-      try _introSort(within: startIndex..<endIndex, by: areInIncreasingOrder)
+      try _stableSortImpl(by: areInIncreasingOrder)
     }
   }
 }
 
 extension MutableCollection where Self: BidirectionalCollection {
+  /// Sorts `self[range]` according to `areInIncreasingOrder`. Stable.
+  ///
+  /// - Precondition: `sortedEnd != range.lowerBound`
+  /// - Precondition: `elements[..<sortedEnd]` are already in order.
   @inlinable
   internal mutating func _insertionSort(
-    within range: Range<Index>, 
+    within range: Range<Index>,
+    sortedEnd: Index,
     by areInIncreasingOrder: (Element, Element) throws -> Bool
   ) rethrows {
-
-    guard !range.isEmpty else { return }
-
-    let start = range.lowerBound
-
-    // Keep track of the end of the initial sequence of sorted elements. One
-    // element is trivially already-sorted, thus pre-increment Continue until
-    // the sorted elements cover the whole sequence
-    var sortedEnd = index(after: start)
-
+    var sortedEnd = sortedEnd
+    
+    // Continue sorting until the sorted elements cover the whole sequence.
     while sortedEnd != range.upperBound {
-      // get the first unsorted element
-      // FIXME: by stashing the element, instead of using indexing and swapAt,
-      // this method won't work for collections of move-only types.
-      let x = self[sortedEnd]
-
-      // Look backwards for x's position in the sorted sequence,
-      // moving elements forward to make room.
       var i = sortedEnd
+      // Look backwards for `self[i]`'s position in the sorted sequence,
+      // moving each element forward to make room.
       repeat {
         let j = index(before: i)
-        let predecessor = self[j]
-
-        // If closure throws, put the element at right place and rethrow.
-        do {
-          // if x doesn't belong before y, we've found its position
-          if try !areInIncreasingOrder(x, predecessor) {
-            break
-          }
-        } catch {
-          self[i] = x
-          throw error
+        
+        // If `self[i]` doesn't belong before `self[j]`, we've found
+        // its position.
+        if try !areInIncreasingOrder(self[i], self[j]) {
+          break
         }
-
-        // Move y forward
-        self[i] = predecessor
+        
+        // Swap the elements at `i` and `j`.
```

Comment:
Nice comment =)

Reply:
😅 I think I was just excited about not using the temporary any more